Template Pipeline: finish support for host bindings

dylhunn opened the pull request:
This PR fixes the remaining tests in the host bindings directory. Some miscellaneous surprises may remain in other directories.

See individual commits for details.

dylhunn committed:
refactor(compiler): Refactor the compilation job and unit abstractions

Convert `CompilationJob` into a abstract class in order to extract common code. Separate host binding jobs and units, in order to allow for more code sharing.

dylhunn committed:
refactor(compiler): Extract host binding static attributes to `hostAttrs`

Host bindings can apply static attributes. These will be extracted to a `hostAttrs` field on the host binding function's metadata.

In order to achieve this, we add an `attributes` field to the host binding job. Then, we peform attribute exraction on host bindings. We finally populate the `attributes` field directly, instead of relying on a `consts` array.

dylhunn committed:
refactor(compiler): Support synthetic host property bindings.

Host property bindings can be animation bindings, and should be ingested and emitted as such, as well as being processed by the renaming phase.

dylhunn committed:
refactor(compiler): Backwards compatibility with TDB for host attributes

TemplateDefinitionBuilder only extracts host attributes if they are text attributes. For example, `[attr.foo]="'my-value'"` is not extracted despite being a string literal, because it is not a text attribute.

dylhunn committed:
refactor(compiler): Support class and style attrs in host bindings

The template pipeline is already capable of parsing and processing class and style attributes on templates. We now extend that functionality to host bindings.

The parser, for some reason, splits out class and style attributes into a `specialAttributes` field. We merge them back into the main attributes map, and allow the template pipeline to process them normally.

dylhunn committed:
refactor(compiler): More closely match TDB binding ordering

For host bindings, `TemplateDefinitionBuilder` seems to use a different binding ordering, in which style bindings come after all the property bindings. We approximate that by treating `hostProperty` differently from `property` in the ordering phase.

dylhunn committed:
refactor(compiler): Support host binding listeners

We can now ingest host bindings listeners into the template pipeline, and process them using the pre-existing phases.

dylhunn committed:
refactor(compiler): Support host animation events.

Animation listeners on host bindings result in a special `syntheticHostListener` instruction. We can now emit this instruction.

Additionally, the naming phase for events has been slightly refactored to smoothly incorporate whether the event is from a host listener, as well as whether it is an animation listener.

dylhunn committed:
refactor(compiler): Generalize the ordering phase to also order create mode

`syntheticHostListener` and `listener` have ordering dependencies. We reuse the existing ordering phase, and generalize it to also order create mode instructions.

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/const_collection.ts:
```diff
@@ -31,19 +32,36 @@ export function phaseConstCollection(cpl: ComponentCompilationJob): void {
+            if (attrArray.entries.length > 0) {
+              op.attributes = job.addConst(attrArray);
+            }
+            elementAttributes.delete(op.xref);
```
Is this intended to affect the entries present in the `else if` block below? (That's what I'm assuming from the comment "any remaining element attributes ..."). If so I'm not sure it'll work properly because the map is repopulated each time this function is called

dylhunn replied:
Nice catch! This is indeed nonsense, left over from some rewriting. The two cases are completely separate, and the host attributes go in a different place anyway, so there's no need to delete.

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/parse_extracted_styles.ts:
```diff
@@ -9,15 +9,15 @@
-export function phaseParseExtractedStyles(cpl: ComponentCompilationJob) {
-  for (const [_, view] of cpl.views) {
-    for (const op of view.create) {
+export function phaseParseExtractedStyles(cpl: CompilationJob) {
```
nit: cpl => job

dylhunn replied:
Fixed in a subsequent commit, which renames all these arguments in all phases.

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/ordering.ts:
```diff
@@ -0,0 +1,113 @@
+
+export function phaseOrdering(job: CompilationJob) {
+  for (const unit of job.units) {
+    const rules = unit.create.head ? CREATE_ORDERING : UPDATE_ORDERING;
```
Seems unused

dylhunn replied:
Indeed, removed.

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/ordering.ts:
```diff
@@ -0,0 +1,113 @@
+    // still have ops pulled at the end, put them back in the correct order.
+
+    // Create mode:
+    let opsToOrder = [];
```
These loops are pretty similar, you could probably extract them into a helper function if you want, and just do: `helper(unit.create, CREATE_ORDERING); helper(unit.update, UPDATE_ORDERING)`

dylhunn replied:
Unfortunately, it's a huge pain to get the types working correctly here, since create and update ops don't share a supertype. I have done my best to make the repeated code as compact as possible, but we should consider fixing this as part of a bigger refactor of Op.

dylhunn committed:
refactor(compiler): Support `$event` in host bindings

Run the existing phase that deals with `$event` during host binding compilation.

dylhunn committed:
refactor(compiler): Enable additional passing tests

Some additional tests were already passing.

dylhunn committed:
refactor(compiler): Share the same phase list for components and hosts

Component compilation and host binding compilation previously used separate compilation emit functions. This was a bit messy, because we had to manage the relative orders of both phase lists. Indeed, there was already some inconsistency between the precise orders!

This commit refactors the emit functions to share the same phase list, and thus guarantees they will always be in the same order.

dylhunn commented:
caretaker: no impact on g3

jessicajaniuk commented:
This PR was merged into the repository by commit c2d859241b5c55af4f7d34551ab88331232b9caf.
